### PR TITLE
[*] Retrocompat - Fix lumber generation on MySql 5.6.

### PR DESCRIPTION
**Pull request by slimee**

### CODE REVIEW

**Comment by slimee** [comment]
[(?) P1 - Lumber - Generate command fails on a MySQL v5.6 database](https://trello.com/c/r3SxgTi4/3988-p1-lumber-generate-command-fails-on-a-mysql-v56-database)

File: services/analyzer/mysql-table-constraints-getter.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        FROM information_schema.table_constraints AS tableConstraints
+        JOIN information_schema.key_column_usage AS keyColumnUsage
+          ON tableConstraints.table_name = keyColumnUsage.table_name
+          AND tableConstraints.constraint_name = keyColumnUsage.constraint_name
+        LEFT JOIN information_schema.STATISTICS AS uniqueIndexes
+          ON keyColumnUsage.column_name = uniqueIndexes.column_name
+          AND tableConstraints.constraint_name = uniqueIndexes.INDEX_NAME
+        WHERE tableConstraints.table_schema = :schemaName
+          AND tableConstraints.table_name = :table
+        ORDER BY uniqueIndexes.SEQ_IN_INDEX;
```
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: Nice ! way more readable 👍

File: services/analyzer/mysql-table-constraints-getter.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    `;
+
+    const constraints = (await queryInterface.sequelize
+      .query(query, { type: queryInterface.sequelize.QueryTypes.SELECT, replacements }))
+      .map(({ sequenceInIndex, ...constraint }) => constraint);
```
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: [Readability] Can you add a little notice about what this line does please ?

File: services/analyzer/mysql-table-constraints-getter.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+function MysqlTableConstraintsGetter(databaseConnection) {
+  const queryInterface = databaseConnection.getQueryInterface();
+
+  this.convertToUniqueIndexArray = (constraints) => {
+    const uniqueIndexes = {};
+    constraints.filter((constraint) => constraint.columnType === 'UNIQUE')
+      .forEach((item) => {
```
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: The data we are working on there are quite hard to follow. May I suggest this naming instead for readability ?
```suggestion
      .forEach((uniqueConstraint) => {
```

File: services/analyzer/mysql-table-constraints-getter.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+function MysqlTableConstraintsGetter(databaseConnection) {
+  const queryInterface = databaseConnection.getQueryInterface();
+
+  this.convertToUniqueIndexArray = (constraints) => {
```
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: A little suggestion for readability
```suggestion
  this.regroupUniqueConstraintIntoArray = (constraints) => {
```

File: services/analyzer/mysql-table-constraints-getter.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    constraints.filter((constraint) => constraint.columnType === 'UNIQUE')
+      .forEach((item) => {
+        uniqueIndexes[item.constraintName] = uniqueIndexes[item.constraintName] || [];
+        uniqueIndexes[item.constraintName].push(item.columnName);
+      });
+    const uniqueIndexArray = Object.values(uniqueIndexes);
+    return uniqueIndexArray.length ? uniqueIndexArray : null;
+  };
+
+  this.applyUniqueIndexArray = (constraints) => {
```
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: Again, only a little suggestion
```suggestion
  this.addUniqueConstraintArrayToConstraints = (constraints) => {
```

Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: [Readability] may you add a little notice there to help dev going through ?

File: services/analyzer/mysql-table-constraints-getter.js (hunk 4 above)
Comment (SteveBunlon) [inline services/analyzer/mysql-table-constraints-getter.js]: [Readability] may you add a little notice there to help dev going through ? (Describing the type of data computed there for example)

File: services/analyzer/sequelize-table-constraints-getter.js [hunk 6]
```diff
@@ -57,65 +59,7 @@ function TableConstraintsGetter(databaseConnection, schema) {
...
-            constraintName,
-            tableName,
-            columnType,
-            columnName,
-            foreignTableName,
-            foreignColumnName,
-            uniqueIndexes`;
-        replacements.schemaName = queryInterface.sequelize.config.database;
-        break;
+        return new MysqlTableConstraintsGetter(databaseConnection).perform(table);
```
Comment (SteveBunlon) [inline services/analyzer/sequelize-table-constraints-getter.js]: ⭐️

File: docker-compose.yml [hunk 7]
```diff
@@ -22,9 +22,20 @@ postgres :
...
+  image: mysql:5.6
+  container_name: lumber_sequelize_mysql_5_6
+  environment:
+    MYSQL_ROOT_PASSWORD: secret
+    MYSQL_DATABASE: lumber-sequelize-test
+    MYSQL_USER: forest
+    MYSQL_PASSWORD: secret
+  ports:
+    - "127.0.0.1:8998:3306"
+
```
[the remark this answers is not in the document]
Comment (slimee) [reply docker-compose.yml]: Oh yes you are right, we have to change all this.
Another refacto PR that set clears name is coming just after 👍

File: test/services/mysql-table-constraints-getter.test.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  describe('with no unique index', () => {
+    it('should provide an empty unique constraint array', async () => {
+      expect.assertions(1);
+      const fixture = [
+        { columnType: 'BLABLA', constraintName: 'FOOFOO1' },
+        { columnType: 'BLABLA', constraintName: 'FOOFOO2' },
+        { columnType: 'BLABLA', constraintName: 'FOOFOO3' },
+      ];
+      const tested = new MysqlTableConstraintsGetter(databaseConnectionMock);
+      const actual = tested.convertToUniqueIndexArray(fixture);
```
[the remark this answers is not in the document]
Comment (slimee) [reply test/services/mysql-table-constraints-getter.test.js]: I am ok about the `tested` name, but for actual I like to have it like that.

File: test/services/mysql-table-constraints-getter.test.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      const expected = null;
```
[the remark this answers is not in the document]
Comment (slimee) [reply test/services/mysql-table-constraints-getter.test.js]: Nope, but I like having one blank line before expect :)

File: test/services/mysql-table-constraints-getter.test.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+      expect(actual).toStrictEqual(expected);
```
[the remark this answers is not in the document]
Comment (slimee) [reply test/services/mysql-table-constraints-getter.test.js]: I agree BLABLA and FOOFOO are not welcomed :)

File: services/analyzer/mysql-table-constraints-getter.js (hunk 4 above)
Comment (slimee) [reply services/analyzer/mysql-table-constraints-getter.js]: I don't find `regroup` clearer than `convert`:)

File: services/analyzer/mysql-table-constraints-getter.js (hunk 5 above)
Comment (slimee) [reply services/analyzer/mysql-table-constraints-getter.js]: I prefer using 4 words too, sorry :)

**Comment by SteveBunlon** [comment]
Also I've noted that the constraints are duplicated with the current request. Wouldn't be an issue ? I need to test a little more (end to end) but maybe you have an idea ? : 
Screenshot 2020-01-09 at 18 05 29

**Comment by slimee** [comment]
> Also I've noted that the constraints are duplicated with the current request. Wouldn't be an issue ? I need to test a little more (end to end) but maybe you have an idea ? :
> Screenshot 2020-01-09 at 18 05 29

Wow, that's strange to have that. 
It would be nice to have the create table because I didn't reproduce it on my different tables.

**Comment by SteveBunlon** [comment]
> > Also I've noted that the constraints are duplicated with the current request. Wouldn't be an issue ? I need to test a little more (end to end) but maybe you have an idea ? :
> > Screenshot 2020-01-09 at 18 05 29
> 
> Wow, that's strange to have that.
> It would be nice to have the create table because I didn't reproduce it on my different tables.

Not reproducable in any manner.
